File: app/src/main/java/com/stevenfrew/beatprompter/cache/AudioFile.java
package com.stevenfrew.beatprompter.cache;

import android.media.MediaMetadataRetriever;

import com.stevenfrew.beatprompter.BeatPrompterApplication;
import com.stevenfrew.beatprompter.R;
import com.stevenfrew.beatprompter.cloud.CloudDownloadResult;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

public class AudioFile extends CachedCloudFile
{
    public final static String AUDIOFILE_ELEMENT_TAG_NAME="audiofile";

    AudioFile(CloudDownloadResult result) throws InvalidBeatPrompterFileException
    {
        super(result);
        verifyAudioFile();
    }

    AudioFile(Element e)
    {
        super(e);
    }

    public void writeToXML(Document doc, Element parent)
    {
        Element audioFileElement = doc.createElement(AUDIOFILE_ELEMENT_TAG_NAME);
        super.writeToXML(audioFileElement);
        parent.appendChild(audioFileElement);
    }

    private void verifyAudioFile() throws InvalidBeatPrompterFileException
    {
        try
        {
            // Try to read the length of the track. If it fails, it's not an audio file.
            MediaMetadataRetriever mmr = new MediaMetadataRetriever();
            mmr.setDataSource(mFile.getAbsolutePath());
            mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
        }
        catch(Exception e)
        {
            throw new InvalidBeatPrompterFileException(BeatPrompterApplication.getResourceString(R.string.notAnAudioFile, mName));
        }
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/cache/AudioFile.kt
package com.stevenfrew.beatprompter.cache

import org.w3c.dom.Document
import org.w3c.dom.Element

@CacheXmlTag("audiofile")
/**
 * An audio file from our cache of files.
 */
class AudioFile internal constructor(
	cachedFile: CachedFile,
	val duration: Long
) : CachedFile(cachedFile) {
	override fun writeToXML(doc: Document, element: Element) {
		super.writeToXML(doc, element)
		element.setAttribute(AUDIO_FILE_LENGTH_ATTRIBUTE, "$duration")
	}

	companion object {
		private const val AUDIO_FILE_LENGTH_ATTRIBUTE = "audioLengthMilliseconds"

		fun readAudioFileLengthFromAttribute(element: Element?): Long? {
			if (element?.hasAttribute(AUDIO_FILE_LENGTH_ATTRIBUTE) == true) {
				val lengthString = element.getAttribute(AUDIO_FILE_LENGTH_ATTRIBUTE)
				try {
					return lengthString.toLong()
				} catch (numberFormatException: NumberFormatException) {
					// Attribute is garbage, we'll need to actually examine the file.
				}
			}
			return null
		}
	}
}